### PR TITLE
Add a dependency check for cadence imports

### PR DESCRIPTION
**Pull request by janezpodhostnik**
ref: #5858

Add a dependency check to the FVM code.

If enabled the FVM will expect a function with the following signature:
```cadence
access(all) fun checkDependencies(_ dependenciesAddresses: [Address], _ dependenciesNames: [String], _ authorizers: [Address]) { }
```
The FVM will call this function at the end of the transaction (after fee deduction). It will inject all dependencies of the transaction and all the authorizers of the transaction + the payer. The call of the function is not metered.

If the function is not there the FVM will quietly continue (It will log the problem). 

If the function panics the transaction will fail (the panic message will be in the error). The function can emit events (but they will only be emitted if the transaction doesn't fail).

This is a v0.33 only feature

### CODE REVIEW

**Comment by bluesign** [comment]
what is the use case for this ?

**Review by sideninja — APPROVED** [review]

**Review by turbolent — COMMENTED** [review]
Is the idea to consult the contract update staging contract and check that for all dependencies a contract has been staged?

File: fvm/environment/system_contracts.go [hunk 1]
```diff
@@ -279,3 +279,48 @@ func (sys *SystemContracts) AccountsStorageCapacity(
...
+				),
+				sema.NewVariableSizedType(
+					nil,
+					sema.StringType,
+				),
+				sema.NewVariableSizedType(
+					nil,
+					&sema.AddressType{},
+				),
+			},
```
Comment (turbolent) [inline fvm/environment/system_contracts.go:318]: Maybe extract that, so it doesn't have to be re-allocated each time

**Comment by janezpodhostnik** [comment]
> Is the idea to consult the contract update staging contract and check that for all dependencies a contract has been staged?

yes exactly.

That part will be implemented in the smart contract, which gives us greater flexibility and means that we do not have to use HCU to deploy changes.

**Pull request merged**